src/util/expr_iterator.h

```diff
@@ -0,0 +1,242 @@
+/*******************************************************************\
+
+ Module: exprt iterator module
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_EXPR_ITERATOR_H
+#define CPROVER_UTIL_EXPR_ITERATOR_H
+
+#include <iterator>
+#include <functional>
+#include <set>
+#include <algorithm>
+#include "expr.h"
+#include "invariant.h"
+
+// Forward declarations - table of contents
+
+/// \file Forward depth-first search iterators
+/// These iterators' copy operations are expensive, so use auto&, and avoid
+/// std::next(), std::prev() etc.
+///
+/// Incrementable only by calling ++it (to avoid accidental copying, as
+/// this iterator is not cheap)
+///
+/// Non-const iterators dereference to const exprt (for use with STL
+/// algorithms) but have an extra .mutate() method. That method is used
+/// to access non-const exprt reference but is an expensive operation
+
+/// Naive depth-first-search iterator with .mutate method for
+/// modifying exprt under the iterator.
+class depth_iteratort;
+/// Naive depth-first-search iterator.
+class const_depth_iteratort;
+/// Depth first-search iterator with duplicate avoidance (skips over
+/// duplicates, shows only first encountered child)
+/// Slower iteration than naive version. Very expensive copy.
+class const_unique_depth_iteratort;
+
+/// Depth first search iterator base - iterates over supplied expression
+/// and all its operands recursively.
+/// Base class using CRTP
+/// Do not create stand-alone instances of or upcast to this base
+/// class. It's only purpose is to provide common functions for other
+/// DFS iterators
+/// Do override .push_expr method of this class, but when doing so
+/// make this class a friend so it has acess to that overriden .push_expr
+/// method in the child class
+template<typename depth_iterator_t>
+class depth_iterator_baset:
+public std::iterator<std::forward_iterator_tag, const exprt, void>
+{
+public:
+  bool operator==(const depth_iterator_t &other) const
+  {
+    return m_stack.size()==other.m_stack.size() &&
```

Comment:
It's not generated, because of manually specified constructor (so because of VC).